chainer/iterators/multiprocess_iterator.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import division
+import multiprocessing
+
+import numpy
+import six
+
+from chainer.dataset import iterator
+
+
+class MultiprocessIterator(iterator.Iterator):
+
+    """Dataset iterator that loads examples in parallel.
+
+    This is an implementation of :class:`~chainer.dataset.Iterator` that loads
+    examples with worker processes. It uses the standard :mod:`multiprocessing`
+    module to parallelize the loading. In order to access the dataset by worker
+    processes, the dataset is sent to them in the standard way using pickle.
+
+    Note that this iterator effectively prefetches the examples for the next
+    batch asynchronously after the current batch is returned.
+
+    This iterator shuffles the order of examples at the beginning of each epoch
+    like :class:`ShuffledIterator`.
+
+    Args:
+        dataset (~chainer.dataset.Dataset): Dataset to iterate.
+        batch_size (int): Number of examples within each batch.
+        repeat (bool): If True, then it infinitely loops over the dataset.
+            Otherwise, it stops iteration at the end of the first epoch.
+        n_processes (int): Number of worker processes. The number of CPUs is
+            used by default.
+
+    """
+    def __init__(self, dataset, batch_size, repeat=True, n_processes=None):
+        self.dataset = dataset
+        self.batch_size = batch_size
+        self._repeat = repeat
+        self._order = numpy.random.permutation(len(dataset))
+        self._prefetch_order = None  # used at the end of each epoch
+
+        self.current_position = 0
+        self.epoch = 0
+        self.is_new_epoch = False
+        self._pushed_position = None  # initialized at the first iteration
+
+        self.n_processes = n_processes or multiprocessing.cpu_count()
+
+        self._start = False
+        self._finalized = False
+
+    def __del__(self):
+        self.finalize()
+
+    def __next__(self):
+        if not self._repeat and self.epoch > 0:
+            raise StopIteration
+
+        self.is_new_epoch = False
+        if not self._start:
+            self._init()  # start workers
+            self._invoke_prefetch()  # load for the first iteration
+        batch = self._get()
+        self._invoke_prefetch()  # prefetch for the next iteration
+        return batch
+
+    next = __next__
+
+    @property
+    def epoch_detail(self):
+        return self.epoch + self.current_position / len(self.dataset)
+
+    def finalize(self):
+        if not self._start or self._finalized:
+            return
+
+        self._finalized = True
+        workers = self._workers
+        self._workers = []
```

Comment:
I would like to confirm my understanding. Is it users' code, not this class itself that tries to reuse this iterator. I thought so because this class accesses `_workers` attribute only in `_init` other than `finalize`. But as `_workers` are initialized in `_init`, too, this is not harmful.